minecraft-plugin/src/main/java/com/pvchat/proximityvoicechat/plugin/distancematrix/PlayerDistanceAndVolumeCalculations.java

```diff
@@ -0,0 +1,105 @@
+    }
+
+    //adding payers that hear each other to the volume list
+    private void addToVolumeList(List<PlayerVolumeData> playerVolumeList, List<Player> playersOfSomeWorld) {
+        int size = playersOfSomeWorld.size();
+
+        for (int i = 0; i < size; i++) {
+            for (int j = i + 1; j < size; j++) {
+                //By using for loop in this way(j=i+1) we can avoid checking
+                //volume between two different players more than once which is efficient.
+
+                Player p1 = playersOfSomeWorld.get(i);
+                Player p2 = playersOfSomeWorld.get(j);
+
+                Optional<DiscordUserID> optionalP1DiscordID = discordLink.getDiscordID(p1.getUniqueId());
+                Optional<DiscordUserID> optionalP2DiscordID = discordLink.getDiscordID(p2.getUniqueId());
+                if (optionalP1DiscordID.isEmpty() || optionalP2DiscordID.isEmpty()) continue;
+                //Skiping situations where discord id for any of two players that we check is empty
+
+                int volume;
+                //Calculating volume between two players on the same world
+                if(p1.getWorld() != p2.getWorld()) volume = 0;
+                else volume = calculateVolume(distanceBetweenTwoPlayers(p1.getLocation(), p2.getLocation()));
+
+                //We add players pairs to volume list only if they can hear each other
+                playerVolumeList.add(new PlayerVolumeData(optionalP1DiscordID.get(), optionalP2DiscordID.get(), volume));
+            }
+        }
+    }
+
+    //getting list of player volume pairs
+    public List<PlayerVolumeData> getPlayerVolumeList() {
+        ArrayList<PlayerVolumeData> playerVolumeData = new ArrayList<>();
+//        if (Bukkit.getOnlinePlayers().size() > 1) {
+        Bukkit.getServer().getWorlds().forEach(world -> addToVolumeList(playerVolumeData, new ArrayList<>(Bukkit.getOnlinePlayers())));
+//        }
+        return playerVolumeData;
+    }
+
+    public static List<PlayerVolumeData> filterPlayerVolumeData(List<PlayerVolumeData> matrix, DiscordUserID discordUserID) {
+        return matrix.stream().filter(playerVolumeData -> playerVolumeData.getPlayer1().equals(discordUserID) || playerVolumeData.getPlayer2().equals(discordUserID)).collect(Collectors.toList());
+    }
+
+
+        //Calculating Distance
+    private double distanceBetweenTwoPlayers(Location location1, Location location2) {
+        return location1.distance(location2);
+    }
+
+    //Calculating Volume
+    private int calculateVolume(double distance) {
+        if (distance > maxHearDistance) {
+            return 0;
+        } else if (distance <= noAttenuationDistance) {
+            return 100;
+        } else {
+            return (int) Math.round(100 - ((distance - noAttenuationDistance) * volumeIncreaseFactor));
+        }
+    }
+
+    //updating player list
+    public void updateVolume(Plugin plugin) {
+        Bukkit.getScheduler().runTaskTimer(plugin, () -> {
```

Comment:
By the previous comment I meant to start a discussion about the thread pool used to run this. I did not mean that we should change it from async task to a blocking one. IMO async was better.